test/SparqlAntlrParserTest.cpp

```diff
@@ -1174,3 +1174,65 @@ TEST(SparqlParser, HavingCondition) {
                         {SparqlFilter::LT, "?predicate", "\"<Z\""});
   expectHavingConditionFails("(LANG(?x) = \"en\")");
 }
+
+TEST(SparqlParser, WhereClause) {
+  auto Triples = [](vector<SparqlTriple>&& triples) {
+    return IsTriples(triples);
+  };
+  auto Bind = [](const string& target, const string& expression) {
+    return IsBindd(target, expression);
+  };
+  auto Filters = [](vector<SparqlFilter>&& triples) {
+    return IsFilters(triples);
+  };
+  auto Optional = IsOptional();
+  auto NotOptional = testing::Not(IsOptional());
+  auto HasChildren = [](auto&&... childMatchers) {
+    return IsChildren(vector{childMatchers...});
+  };
+
+  auto expectGraphPattern = [](const string&& input, const auto&... matchers) {
+    expectCompleteParse(parseWhereClause(input,
```

Comment:
I see your point. But there may be nested `GraphPatterns`. So it would be nice to be able to nest this. I will implement a Matcher instead that models this pooling of the testing of the optional attribute, filters and children.